File: TypeScript/src/board.ts
import {Result, ok, err} from "neverthrow";
import { OutOfBounds, TileOccupied } from "./errors";
        
// Types and Interfaces
type StatusMessage = "win" | "draw" | "loss" | "undeclared";
export type Board2D = string[][];
export type Coordinates = [number, number];
export type GamePiece = "X" | "O";

export interface BoardStatus {
    board: Board2D;
    state: StatusMessage;
};

export interface Diagonal {
    left: string[];
    right: string[];
};


export class Board {
    protected _board: Board2D;
    protected _size: number;
    
    constructor(sizeOfBoard = 3) {
        this._size = sizeOfBoard;
        this._board = this._new();
    }

    protected _new(): Board2D {
        return Array(this._size).fill([]).map(() => Array(this._size).fill("#"));
    }; 

    protected _getAllPossibleCombinations(board: Board2D = this._board): string[][] {
        let combinations: string[][] = [...this._board];
        let diagonal: Diagonal = { left: [], right: [] };

        for(let i = 0; i < this._size; i++){
            diagonal.left.push(board[i][i]);
            diagonal.right.push(board[i][this._size-i-1]);
            // Vertical
            combinations.push(board.map((val: string[]) => val[i]))
        };

        return [...combinations, diagonal.left, diagonal.right]
    }

    protected _determineBoardStatus(board: Board2D = this._board): StatusMessage {
        let combinations = this._getAllPossibleCombinations(board);

        let getWinner = (comb: string[], gamePiece: "X" | "O"): Result<{}, {}> => {
            if(comb.filter(c => c === gamePiece).length === 3) return ok({});
            else return err({}); 
        } 
        for (let comb of combinations) {
            if(getWinner(comb, "X").isOk()) return "win";
            if(getWinner(comb, "O").isOk()) return "loss";
        };

        let flattenedCombinations = combinations.reduce((acc, val) => acc.concat(val), []);
        if(!flattenedCombinations.includes("#")) return "draw";

        return "undeclared";
    }

    public placeGamePiece(coords: Coordinates, gamePiece: GamePiece, bot = false): Result<string, OutOfBounds | TileOccupied> {
        let x = coords[0];
        let y = coords[1];
        if(bot) {
            x+=1
            y+=1
        }
        if(x < 1 || x > this._size || y < 1 || y > this._size) {
            return err({
                type: "OutOfBounds",
                minVal: 1,
                maxVal: this._size,
                message: "Coords are out of bounds"
            });
        } else if(this._board[y-1][x-1] === "X" || this._board[y-1][x-1] === "O"){
            return err({
                type: "TileOccupied",
                message: "Tile is occupied"
            });
        };

        this._board[y-1][x-1] = gamePiece;
        return ok("Success");
    }

    public status(): BoardStatus {
        return {board: this._board, state: this._determineBoardStatus()}
    };

    public getBoard(): Board2D {
        return this._board;
    };

};

File: src/board.rs
pub type Coordinates = (usize, usize);
pub type Board2d = [[char; 3]; 3];

#[derive(PartialEq)]
pub enum BoardStatus {
    Win,
    Lose,
    Draw,
    Undeclared
}

pub enum GamePiece {
    X,
    O
}

impl GamePiece {
    pub fn get_char(&self) -> char {
        match self {
            GamePiece::X => return 'X',
            GamePiece::O => return 'O',
        };
    }
}

pub struct Board {
    board: Board2d,
}

impl Board {
    pub fn new() -> Board {
        Board {
            board: [['#'; 3]; 3],
        }
    }
    
    pub fn from(template_board: Board2d) -> Board {
        Board {
            board: template_board,
        }
    }

    pub fn all_combinations(&self) -> Vec<[char; 3]> {
        let mut combinations: Vec<[char; 3]> = Vec::new();
        let mut vertical: [char; 3] = ['#'; 3];

        for row in self.board.iter() {
            combinations.push(row.to_owned());
        }

        for i in 0..=2 {
            for j in 0..=2 {
                vertical[j] = self.board[j][i];
            }

            combinations.push(vertical);
        }

        for i in 0..=2 {
            vertical[i] = self.board[i][i];
        }
        combinations.push(vertical);

        for i in 0..=2 {
            vertical[i] = self.board[i][2 - i];
        }
        combinations.push(vertical);

        combinations
    }

    fn get_winner(combinations: [char; 3], game_piece: GamePiece) -> BoardStatus {
        let game_piece = game_piece.get_char();
        let filtered_combination: Vec<&char> = combinations
            .iter()
            .filter(|comb| **comb == game_piece)
            .collect();

        if filtered_combination.len() == 3 {
            return BoardStatus::Win
        }
        
        BoardStatus::Undeclared
    }

    pub fn status(&self) -> BoardStatus {
        let combinations = self.all_combinations();
        let mut full_board = true;

        for comb in combinations {
            if Board::get_winner(comb, GamePiece::X) == BoardStatus::Win { return BoardStatus::Win }
            if Board::get_winner(comb, GamePiece::O) == BoardStatus::Win { return BoardStatus::Lose }
        
            for tile in comb.iter() {
                if *tile == '#' { full_board = false }
            }      
        }

        if full_board { return BoardStatus::Draw }

        BoardStatus::Undeclared
    }

    pub fn place_game_piece(&mut self, coords: Coordinates, game_piece: GamePiece){
        self.board[coords.1][coords.0] = game_piece.get_char();
    }

    pub fn get_board(&mut self) -> Board2d {
        return self.board; 
    }
}
